File: avivamymoney.js
/**
 * `AVIVA_USER=myuser AVIVA_PWD=mypassword node aviva.js`
 *
 */
const puppeteer = require('puppeteer-extra');

const StealthPlugin = require('puppeteer-extra-plugin-stealth')
puppeteer.use(StealthPlugin())

puppeteer.launch({ headless: false }).then(async browser => {
	
  const page = await browser.newPage();

  try {
    const session = await page.target().createCDPSession();
    const { windowId } = await session.send('Browser.getWindowForTarget');
    //await session.send('Browser.setWindowBounds', {windowId, bounds: {windowState: 'minimized'}});

    await page.setViewport({ width: 768, height: 768 })

    await page.goto('https://www.avivamymoney.co.uk/Login');
    await page.waitForSelector('#onetrust-accept-btn-handler');
    await page.click('#onetrust-accept-btn-handler');

    await page.waitForSelector('a.btn-primary:nth-child(1)');
    await page.type('#Username', process.env.AVIVAMYMONEY_USER, {delay: 10});
    await page.type('#Password', process.env.AVIVAMYMONEY_PWD, {delay: 10});

    await page.waitForTimeout(1000);
    await page.click('a.btn-primary:nth-child(1)');

    await page.waitForSelector("html");

    await page.waitForSelector("#FirstLetter");

    await page.select("#FirstLetter", process.env.AVIVAMYMONEY_WORD.charAt(await page.evaluate('document.querySelector("#FirstElement_Index").getAttribute("value")') - 1));
    await page.select("#SecondLetter", process.env.AVIVAMYMONEY_WORD.charAt(await page.evaluate('document.querySelector("#SecondElement_Index").getAttribute("value")') - 1));
    await page.select("#ThirdLetter", process.env.AVIVAMYMONEY_WORD.charAt(await page.evaluate('document.querySelector("#ThirdElement_Index").getAttribute("value")') - 1));

    await page.click('[name=Next]');
    const element = await page.waitForSelector(".vspace-reset.text-size-42");

    const balance = await element.evaluate(el => el.textContent);
    console.log(balance.replace("£", "").replace(",", ""));

  } catch (error) {
    console.error(error);
    await page.screenshot({ path: 'debug-avivamymoney-' + process.env.AVIVAMYMONEY_USER + '.png' });
  }

  await browser.close();
})

File: avivamymoney.py
#!/usr/bin/env python3
#
# selenium script to get aviva balance
#
# AVIVA_USER=user AVIVA_PWD=password ./aviva.py
#

import time
import re 
import os
import undetected_chromedriver as uc
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support.ui import Select

# Note: use_subprocess=False fails on mac, should be set true
# Note: drive is assumed to be x86-64 on linux, so we copy locally installed chromedriver
#
driver = uc.Chrome(headless=True,use_subprocess=False,driver_executable_path=os.environ['HOME']+"/.local/share/undetected_chromedriver/chromedriver_copy")

try:
    driver.get('https://www.avivamymoney.co.uk/Login')

    wait = WebDriverWait(driver, 10)
    element = wait.until(EC.element_to_be_clickable((By.CSS_SELECTOR,'#onetrust-accept-btn-handler')))
    element.click()

    wait.until(EC.element_to_be_clickable((By.CSS_SELECTOR,'a.btn-primary:nth-child(1)')))

    driver.find_element(By.CSS_SELECTOR,'#Username').send_keys(os.environ['AVIVAMYMONEY_USER'])
    driver.find_element(By.CSS_SELECTOR,'#Password').send_keys(os.environ['AVIVAMYMONEY_PWD'])
    driver.find_element(By.CSS_SELECTOR,'a.btn-primary:nth-child(1)').click()

    wait.until(EC.element_to_be_clickable((By.CSS_SELECTOR,'#FirstLetter')))

    Select(driver.find_element(By.CSS_SELECTOR,'#FirstLetter')).select_by_visible_text(os.environ['AVIVAMYMONEY_WORD'][int(driver.find_element(By.CSS_SELECTOR,'#FirstElement_Index').get_attribute('value'))-1])
    Select(driver.find_element(By.CSS_SELECTOR,'#SecondLetter')).select_by_visible_text(os.environ['AVIVAMYMONEY_WORD'][int(driver.find_element(By.CSS_SELECTOR,'#SecondElement_Index').get_attribute('value'))-1])
    Select(driver.find_element(By.CSS_SELECTOR,'#ThirdLetter')).select_by_visible_text(os.environ['AVIVAMYMONEY_WORD'][int(driver.find_element(By.CSS_SELECTOR,'#ThirdElement_Index').get_attribute('value'))-1])
    driver.find_element(By.CSS_SELECTOR,'[name=Next]').click()

    wait.until(EC.element_to_be_clickable((By.CSS_SELECTOR,'.vspace-reset.text-size-42')))

    print(driver.find_element(By.CSS_SELECTOR,'.vspace-reset.text-size-42').get_attribute('innerHTML').replace(',', '').replace('£', ''))

except:
    driver.save_screenshot('debug-avivamymoney-'+os.environ['AVIVAMYMONEY_USER']+'.png')
